src/intptrcast.rs

```diff
@@ -72,12 +74,45 @@ impl<'mir, 'tcx> GlobalStateInner {
                     None
                 }
             }
-        };
-        // Pointers created from integers are untagged.
-        Pointer::new(
-            alloc_id.map(|alloc_id| Tag { alloc_id, sb: SbTag::Untagged }),
-            Size::from_bytes(addr),
-        )
+        }?;
+
+        match global_state.provenance_mode {
+            ProvenanceMode::Legacy => Some(alloc_id),
+            ProvenanceMode::Permissive if global_state.exposed.contains(&alloc_id) =>
+                Some(alloc_id),
+            _ => None,
+        }
+    }
+
+    pub fn expose_addr(ecx: &MiriEvalContext<'mir, 'tcx>, alloc_id: AllocId) {
+        trace!("Exposing allocation id {:?}", alloc_id);
+
+        let mut global_state = ecx.machine.intptrcast.borrow_mut();
+        global_state.exposed.insert(alloc_id);
+    }
+
+    pub fn ptr_from_addr(ecx: &MiriEvalContext<'mir, 'tcx>, addr: u64) -> Pointer<Option<Tag>> {
+        trace!("Transmuting 0x{:x} to a pointer", addr);
+
+        // TODO: fix this at some point once we deal with function pointers
+        // Pointer::new(None, Size::from_bytes(addr))
+        Self::ptr_from_casted_addr(ecx, addr)
+    }
+
+    pub fn ptr_from_casted_addr(
```

Comment:
I am very confused by your code structure here.
In the `machine` we have `ptr_from_addr_{cast,transmute}`, both of which call `ptr_from_addr`, which has a comment talking about "transmute", and then calls `ptr_from_casted_addr`...?

Please have both `ptr_from_addr_{cast,transmute}` here in intptrcast.rs, and call the corresponding function from the machine.